src/adc.rs

```diff
@@ -0,0 +1,562 @@
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+))]
+use crate::gpio::{gpiod, gpioe, gpiof};
+
+#[cfg(feature = "stm32f303")]
+use crate::stm32::{ADC1, ADC1_2, ADC2};
+
+#[cfg(any(
+    feature = "stm32f303xb",
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+))]
+use crate::stm32::{ADC3, ADC3_4, ADC4};
+
+/// ADC configuration
+///
+/// TODO: Remove `pub` from the register block once all functionalities are implemented.
+/// Leave it here until then as it allows easy access to the registers.
+pub struct Adc<ADC> {
+    pub rb: ADC,
+    clocks: Clocks,
+    ckmode: CKMODE,
+    operation_mode: Option<OperationMode>,
+}
+
+/// ADC sampling time
+///
+/// Each channel can be sampled with a different sample time.
+/// For Sampletime T_n the total conversion time (in ADC clock cycles) is
+/// 12.5 + (n + .5)
+///
+/// TODO: there are boundaries on how this can be set depending on the hardware.
+/// Check them and implement a sample time setting mechanism.
+pub enum SampleTime {
+    T_1,
+    T_2,
+    T_4,
+    T_7,
+    T_19,
+    T_61,
+    T_181,
+    T_601,
+}
+
+impl Default for SampleTime {
+    fn default() -> Self {
+        SampleTime::T_19
+    }
+}
+
+impl SampleTime {
+    /// Conversion to bits for SMP
+    fn bitcode(&self) -> u8 {
+        match self {
+            SampleTime::T_1 => 0b000,
+            SampleTime::T_2 => 0b001,
+            SampleTime::T_4 => 0b010,
+            SampleTime::T_7 => 0b011,
+            SampleTime::T_19 => 0b100,
+            SampleTime::T_61 => 0b101,
+            SampleTime::T_181 => 0b110,
+            SampleTime::T_601 => 0b111,
+        }
+    }
+}
+
+#[derive(Clone, Copy, PartialEq)]
+/// ADC operation mode
+///
+/// TODO: Implement other modes (DMA, Differential,…)
+pub enum OperationMode {
+    OneShot,
+}
+
+#[derive(Clone, Copy, PartialEq)]
+/// ADC CKMODE
+///
+/// TODO: Add ASYNCHRONOUS mode
+pub enum CKMODE {
+    // ASYNCHRONOUS = 0,
+    SYNCDIV1 = 1,
+    SYNCDIV2 = 2,
+    SYNCDIV4 = 4,
+}
+
+impl Default for CKMODE {
+    fn default() -> Self {
+        CKMODE::SYNCDIV2
+    }
+}
+
+// ADC3_2 returns a pointer to a adc1_2 type, so this from is ok for both.
+#[cfg(feature = "stm32f303")]
+impl From<CKMODE> for stm32f3::stm32f303::adc1_2::ccr::CKMODE_A {
+    fn from(ckmode: CKMODE) -> Self {
+        match ckmode {
+            //CKMODE::ASYNCHRONOUS => CKMODE_A::ASYNCHRONOUS,
+            CKMODE::SYNCDIV1 => CKMODE_A::SYNCDIV1,
+            CKMODE::SYNCDIV2 => CKMODE_A::SYNCDIV2,
+            CKMODE::SYNCDIV4 => CKMODE_A::SYNCDIV4,
+        }
+    }
+}
+
+/// ADC data register alignment
+pub enum Align {
+    /// Right alignment of output data
+    Right,
+    /// Left alignment of output data
+    Left,
+}
+
+impl Default for Align {
+    fn default() -> Self {
+        Align::Right
+    }
+}
+
+impl Align {
+    /// Conversion to bits for ALIGN in ADCx_CFGR
+    fn bitvalue(&self) -> bool {
```

Comment:
Consider replacing this with an `impl From<Align> for ALIGN_A`, like you did with CKMODE.